### PR TITLE
(MBA-6): Implement Register, Log in & Password reset endpoints

### PR DESCRIPTION
**Pull request by Qildeli**
Implemented user authentication system, including:

- User Registration, Login, Password Reset and Password Change endpoints.
- JWT Access and Refresh Token integration for secure login.
- JWTAuthentication class to enable JWT in Swagger documentation.
- Full implementation of views, serializers, and URLs for all authentication endpoints.

I'll provide tests later <3.

### CODE REVIEW

File: src/apps/users/jwt_utils.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from datetime import datetime, timedelta
```
Comment (TatoSoselia) [inline src/apps/users/jwt_utils.py:6]: Missing module docstring

File: src/apps/users/jwt_utils.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import jwt
+from django.conf import settings
+
+
+def generate_access_token(user_id):
```
Comment (TatoSoselia) [inline src/apps/users/jwt_utils.py:12]: Missing function docstring

File: src/apps/users/jwt_utils.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        "user_id": str(user_id),
+        "exp": datetime.timestamp(datetime.now() + timedelta(minutes=5)),
+    }
+    access_token = jwt.encode(
+        access_token_payload, settings.SECRET_KEY, algorithm="HS256"
+    )
+    return access_token
+
+
+def generate_refresh_token(user_id):
```
Comment (TatoSoselia) [inline src/apps/users/jwt_utils.py:27]: Missing function docstring

File: src/apps/users/jwt_utils.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        "user_id": str(user_id),
+        "exp": datetime.timestamp(datetime.now() + timedelta(minutes=30)),
+    }
+    refresh_token = jwt.encode(
+        refresh_token_payload, settings.SECRET_KEY, algorithm="HS256"
+    )
+    return refresh_token
+
+
+def decode_refresh_token(refresh_token):
```
Comment (TatoSoselia) [inline src/apps/users/jwt_utils.py:42]: Missing function docstring

File: src/apps/users/serializers.py [hunk 5]
```diff
@@ -1,10 +1,57 @@
...
+
+class PasswordResetSerializer(serializers.Serializer):
+    reset_token = serializers.CharField()
+    new_password = serializers.CharField(
+        style={
+            "input_type": "password",
+        },
+        write_only=True,
+    )
+    new_password2 = serializers.CharField(
```
Comment (TatoSoselia) [inline src/apps/users/serializers.py]: use new_password_confirm instend of 'new_password2'

File: src/apps/users/views.py [hunk 6]
```diff
@@ -1,3 +1,152 @@
...
+    serializer_class = RefreshTokenSerializer
+    permission_classes = [AllowAny]
+
+    def post(self, request):
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        try:
+            access_token = update_access_token(serializer.data.get("refresh_token"))
+            return Response({"access_token": access_token}, status=status.HTTP_200_OK)
```
Comment (TatoSoselia) [inline src/apps/users/views.py]: take out return statment from try except block.

File: src/apps/users/views.py [hunk 7]
```diff
@@ -1,3 +1,152 @@
...
+    def post(self, request):
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        email = serializer.validated_data.get("email")
+        try:
+            user = get_user_model().objects.get(email=email)
+        except get_user_model().DoesNotExist:
+            return Response(
+                {"error": "User with this email does not exist"},
+                status=status.HTTP_400_BAD_REQUEST,
```
Comment (TatoSoselia) [inline src/apps/users/views.py]: using a 404 status code instead of 400 in this context.

File: src/apps/users/views.py [hunk 8]
```diff
@@ -1,3 +1,152 @@
...
+            )
+
+        reset_token = generate_access_token(user.id)
+
+        # Generate password forget url
+        password_forget_url = (
+            request.build_absolute_uri(reverse("password-forget"))
```
Comment (TatoSoselia) [inline src/apps/users/views.py]: use f formating for full URL

File: src/apps/users/views.py [hunk 9]
```diff
@@ -1,3 +1,152 @@
...
+
+        except jwt.ExpiredSignatureError:
+            return Response(
+                {"error": "Reset token has expired."},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        except (jwt.InvalidTokenError, get_user_model().DoesNotExist):
+            return Response(
+                {"error": "Invalid token or user does not exist."},
+                status=status.HTTP_400_BAD_REQUEST,
```
Comment (TatoSoselia) [inline src/apps/users/views.py]: using a 404 status code instead of 400 in this context.

File: src/apps/users/views.py [hunk 10]
```diff
@@ -1,3 +1,152 @@
...
+            user.save()
+            return Response(
+                {"message": "Password has been reset successfully."},
+                status=status.HTTP_200_OK,
+            )
+
+        except jwt.ExpiredSignatureError:
+            return Response(
+                {"error": "Reset token has expired."},
+                status=status.HTTP_400_BAD_REQUEST,
```
Comment (TatoSoselia) [inline src/apps/users/views.py]: using a 401 status code instead of 400 in this context.

File: src/apps/users/views.py [hunk 11]
```diff
@@ -1,3 +1,152 @@
...
+            user_id = payload.get("user_id")
+            if user_id is None:
+                raise InvalidTokenError("Invalid token")
+
+            user = get_user_model().objects.get(id=user_id)
+            user.set_password(new_password)
+            user.save()
+            return Response(
+                {"message": "Password has been reset successfully."},
+                status=status.HTTP_200_OK,
```
Comment (TatoSoselia) [inline src/apps/users/views.py]: using a 201 status code instead of 200 in this context.

File: src/core/jwt_authentication.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+from rest_framework.authentication import TokenAuthentication
+
+from apps.users.jwt_utils import decode_refresh_token
+
+
+class JWTAuthentication(TokenAuthentication):
+    def authenticate(self, request):
+        # Extract the JWT from the Authorization header
+        jwt_token = request.META.get("HTTP_AUTHORIZATION")
+        print(jwt_token)
```
Comment (TatoSoselia) [inline src/core/jwt_authentication.py]: remove print statement.

File: src/core/jwt_authentication.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+            raise exceptions.AuthenticationFailed("Invalid signature")
+        except jwt.exceptions.DecodeError:
+            raise exceptions.AuthenticationFailed("Invalid token")
+
+        # Get the user from the database
+        user_id = payload.get("user_id")
+        if user_id is None:
+            raise exceptions.AuthenticationFailed("User identifier not found in JWT")
+
+        user = get_user_model().objects.filter(id=user_id).first()
```
Comment (TatoSoselia) [inline src/core/jwt_authentication.py]: method 'first ' is not necessary.

File: src/core/jwt_authentication.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
+import jwt
```
Comment (TatoSoselia) [inline src/core/jwt_authentication.py:6]: Missing docstrings

File: src/core/jwt_authentication.py [hunk 15]
```diff
@@ -27,7 +38,8 @@ def authenticate(self, request):
         if user_id is None:
             raise exceptions.AuthenticationFailed("User identifier not found in JWT")
 
-        user = get_user_model().objects.filter(id=user_id).first()
+        user = get_user_model().objects.get(id=user_id)
+        print(user)
```
Comment (TatoSoselia) [inline src/core/jwt_authentication.py]: remove print function

**Review by TatoSoselia — APPROVED** [review]

**Review by rGudavadze — APPROVED** [review]

**Pull request merged**